provider/metrics/metrics.go

```diff
@@ -0,0 +1,223 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/pkg/errors"
+	"github.com/virtual-kubelet/azure-aci/client/aci"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	stats "github.com/virtual-kubelet/virtual-kubelet/node/api/statsv1alpha1"
+	"github.com/virtual-kubelet/virtual-kubelet/trace"
+	"golang.org/x/sync/errgroup"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	ContainerGroupCacheTTLSeconds = 60 * 5
+)
+
+// package dependency: query the Pods in current virtual nodes. it usually is ResourceManager
+type PodGetter interface {
+	GetPods() []*v1.Pod
+}
+
+// package dependency: query the Pod's correspoinding Container Group metrics from Container Insights
+type ContainerGroupMetricsGetter interface {
+	GetContainerGroupMetrics(ctx context.Context, resourceGroup, containerGroup string, options aci.MetricsRequest) (*aci.ContainerGroupMetricsResult, error)
+}
+
+// package dependency: query the Container Group information
+type ContaienrGroupGetter interface {
+	GetContainerGroup(ctx context.Context, resourceGroup, containerGroupName string) (*aci.ContainerGroup, *int, error)
+}
+
+/*
+there are difference implementation of query Pod's statistics.
+this interface is for mocking in unit test
+*/
+type podStatsGetter interface {
+	getPodStats(ctx context.Context, pod *v1.Pod) (*stats.PodStats, error)
+}
+
+type ACIPodMetricsProvider struct {
+	nodeName           string
+	metricsSync        sync.Mutex
+	podGetter          PodGetter
+	aciCGGetter        ContaienrGroupGetter
+	aciCGMetricsGetter ContainerGroupMetricsGetter
+	podStatsGetter     podStatsGetter
+}
+
+func NewACIPodMetricsProvider(nodeName, aciResourcegroup string, podGetter PodGetter, aciCGGetter ContaienrGroupGetter, aciCGMetricsGetter ContainerGroupMetricsGetter) *ACIPodMetricsProvider {
+	provider := ACIPodMetricsProvider{
+		nodeName:           nodeName,
+		podGetter:          podGetter,
+		aciCGGetter:        aciCGGetter,
+		aciCGMetricsGetter: aciCGMetricsGetter,
+	}
+
+	// containerInsightGetter := WrapCachedPodStatsGetter(
```

Comment:
Deleted commented code

Reply:
good catch. I comment the lines for diagnose a bug. I forget push them back. I have fixed that.